src/run/storage/request-scoped-in-memory-cache.cts

```diff
@@ -0,0 +1,108 @@
+export function setInMemoryCacheMaxSizeFromNextConfig(size: unknown) {
+  if (typeof size === 'number') {
+    extendedGlobalThis[IN_MEMORY_CACHE_MAX_SIZE] = size
+  }
+}
+
+const estimateBlobSize = (valueToStore: BlobType | null | Promise<unknown>): number => {
+  // very approximate size calculation to avoid expensive exact size calculation
+  // inspired by https://github.com/vercel/next.js/blob/ed10f7ed0246fcc763194197eb9beebcbd063162/packages/next/src/server/lib/incremental-cache/file-system-cache.ts#L60-L79
+  if (valueToStore === null || isPromise(valueToStore) || isTagManifest(valueToStore)) {
+    return 25
+  }
+  if (isHtmlBlob(valueToStore)) {
+    return valueToStore.html.length
+  }
+  let knownKindFailed = false
+  try {
+    if (valueToStore.value?.kind === 'FETCH') {
+      return valueToStore.value.data.body.length
+    }
+    if (valueToStore.value?.kind === 'APP_PAGE') {
+      return valueToStore.value.html.length + (valueToStore.value.rscData?.length ?? 0)
+    }
+    if (valueToStore.value?.kind === 'PAGE' || valueToStore.value?.kind === 'PAGES') {
+      return valueToStore.value.html.length + JSON.stringify(valueToStore.value.pageData).length
+    }
+    if (valueToStore.value?.kind === 'ROUTE' || valueToStore.value?.kind === 'APP_ROUTE') {
+      return valueToStore.value.body.length
+    }
+  } catch {
+    // size calculation rely on the shape of the value, so if it's not what we expect, we fallback to JSON.stringify
+    knownKindFailed = true
+  }
+
+  // fallback for not known kinds or known kinds that did fail to calculate size
+  // we should also monitor cases when fallback is used because it's not the most efficient way to calculate/estimate size
+  // and might indicate need to make adjustments or additions to the size calculation
+  recordWarning(
+    new Error(
+      `Blob size calculation did fallback to JSON.stringify. Kind: KnownKindFailed: ${knownKindFailed}, ${valueToStore.value?.kind ?? 'undefined'}`,
+    ),
+  )
+
+  return JSON.stringify(valueToStore).length
+}
+
+function getInMemoryLRUCache() {
+  if (typeof extendedGlobalThis[IN_MEMORY_LRU_CACHE] === 'undefined') {
+    const maxSize =
+      typeof extendedGlobalThis[IN_MEMORY_CACHE_MAX_SIZE] === 'number'
+        ? extendedGlobalThis[IN_MEMORY_CACHE_MAX_SIZE]
+        : DEFAULT_FALLBACK_MAX_SIZE
+
+    extendedGlobalThis[IN_MEMORY_LRU_CACHE] =
+      maxSize === 0
+        ? null // if user sets 0 in their config, we should honor that and not use in-memory cache
+        : new LRUCache<string, BlobType | typeof NullValue | Promise<BlobType | null>>({
+            max: 1000,
+            maxSize,
+            sizeCalculation: (valueToStore) => {
+              return estimateBlobSize(valueToStore === NullValue ? null : valueToStore)
+            },
+          })
+  }
+  return extendedGlobalThis[IN_MEMORY_LRU_CACHE]
+}
+
+interface RequestScopedInMemoryCache {
+  get(key: string): BlobType | null | Promise<BlobType | null> | undefined
```

Comment:
Noting here that we're using `undefined` and `null` as distinctly different values